x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/DoubleScriptFieldScript.java

```diff
@@ -35,55 +35,38 @@
         DoubleScriptFieldScript newInstance(LeafReaderContext ctx) throws IOException;
     }
 
-    private double[] values = new double[1];
-    private int count;
-
     public DoubleScriptFieldScript(Map<String, Object> params, SearchLookup searchLookup, LeafReaderContext ctx) {
         super(params, searchLookup, ctx);
     }
 
+    public abstract double[] execute();
+
     /**
      * Execute the script for the provided {@code docId}.
      */
-    public final void runForDoc(int docId) {
-        count = 0;
+    public final double[] runForDoc(int docId) {
         setDocument(docId);
-        execute();
+        return execute();
     }
 
-    /**
-     * Values from the last time {@link #runForDoc(int)} was called. This array
-     * is mutable and will change with the next call of {@link #runForDoc(int)}.
-     * It is also oversized and will contain garbage at all indices at and
-     * above {@link #count()}.
-     */
-    public final double[] values() {
-        return values;
+    public static double[] convertFromDouble(double v) {
+        return new double[] { v };
     }
 
-    /**
-     * The number of results produced the last time {@link #runForDoc(int)} was called.
-     */
-    public final int count() {
-        return count;
-    }
-
-    private void collectValue(double v) {
-        if (values.length < count + 1) {
-            values = ArrayUtil.grow(values, count + 1);
+    public static double[] convertFromCollection(Collection<?> v) {
+        double[] result = new double[v.size()];
+        int i = 0;
+        for (Object o : v) {
+            result[i++] = ((Number) o).doubleValue();
         }
-        values[count++] = v;
+        return result;
     }
 
-    public static class Value {
-        private final DoubleScriptFieldScript script;
-
-        public Value(DoubleScriptFieldScript script) {
-            this.script = script;
-        }
-
-        public void value(double v) {
-            script.collectValue(v);
+    public static double[] convertFromDef(Object o) {
+        if (o instanceof Double) {
+            return convertFromDouble(((Double) o).doubleValue());
+        } else {
+            return (double[]) o;
```

Comment:
oh boy I was hoping we would not need this sort of stuff, but I guess we do? I mean the instanceof as well as the cast to double array

Reply:
:cry: Me too! I imagine the painless folk are thinking about it, but I'm not sure. This seems like a perfect spot for invokedynamic, but I'm not an expert at all.

Reply:
what happens if we do without this conversion? Really bad I guess?

Reply:
@javanna without it you can't really return `def`. You get a lot of funny class cast exceptions.

@stu-elastic or @jdconrad do we have plans `indy-ify` this or something? So it'd call the converter based on the `def` type. Is that what #61389 is all about?

Reply:
We needed to get this in your hands ASAP.  We'll be improving it: https://github.com/elastic/elasticsearch/issues/61389

Reply:
Indy for this is really challenging. We are going to look into it, but allowing you to do def conversions this way ensures we have something for runtime fields now that doesn't use reflection invocation. I would recommend that this cover all numeric cases including byte through long as well. Check out something like DefMath.plus.

Reply:
Gotcha. So if I implement `convertFromDef` returning a `def` type is all on me at the moment. I have to do what `DefMath` does, basically.

Reply:
@nik9000 Yes, that's correct. Again we hope to rectify this, but wanted to make sure we had something that was usable now.

Reply:
:metal: